kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ProtocolMetadataLogReplay.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.checksum.CRCInfo;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.lang.Lazy;
+import io.delta.kernel.internal.metrics.SnapshotMetrics;
+import io.delta.kernel.internal.snapshot.LogSegment;
+import io.delta.kernel.internal.tablefeatures.TableFeatures;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Static utility class for loading Protocol and Metadata from Delta log files. */
+public class ProtocolMetadataLogReplay {
+
+  private static final Logger logger = LoggerFactory.getLogger(ProtocolMetadataLogReplay.class);
+
+  /** Read schema when searching for the latest Protocol and Metadata. */
+  public static final StructType PROTOCOL_METADATA_READ_SCHEMA =
+      new StructType().add("protocol", Protocol.FULL_SCHEMA).add("metaData", Metadata.FULL_SCHEMA);
+
+  /** Result of loading Protocol and Metadata from a LogSegment. */
+  public static class Result {
+    public final Protocol protocol;
+    public final Metadata metadata;
+    private final long logFilesRead;
+
+    public Result(Protocol protocol, Metadata metadata, long logFilesRead) {
+      this.protocol = protocol;
+      this.metadata = metadata;
+      this.logFilesRead = logFilesRead;
+    }
+
+    public long getLogFilesRead() {
+      return logFilesRead;
+    }
+  }
+
+  /**
+   * Loads the latest Protocol and Metadata from the log files in the given LogSegment.
+   *
+   * <p>Uses the provided lazy CRC loader to bound how many delta files it reads, and to ensure we
+   * only read the CRC file if needed.
+   *
+   * <p>We read delta files in reverse order (newest first) searching for the latest Protocol and
+   * Metadata. When we reach the version just before (greater than) the CRC file version (if
+   * present), we lazily load the CRC file to fill in any missing Protocol or Metadata, avoiding
+   * reading older delta files.
+   *
+   * <p>Also validates that the Kernel can read the table at the loaded protocol version.
+   */
+  public static Result loadProtocolAndMetadata(
```

Comment:
Mostly moved verbatim from LogReplay.java. Changes:
- crc loading is truly lazy now
- clearer logging done in one place only